src/docfx/lib/ProcessUtility.cs

```diff
@@ -326,22 +325,18 @@ public static Task WriteFile(string path, Action<Stream> write)
         public static async Task RunInsideMutex(string mutexName, Func<Task> action)
         {
             Debug.Assert(!string.IsNullOrEmpty(mutexName));
-
-            // avoid the RunInsideMutex to be nested used
-            // doesn't support to require a lock before releasing a lock
-            // which may cause deadlock
-            if (t_innerCall.Value != null)
-            {
-                throw new NotImplementedException("Nested call to RunInsideMutex is not supported yet");
-            }
-
-            t_innerCall.Value = new object();
+            var lockPath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(mutexName));
 
             try
             {
-                Directory.CreateDirectory(AppData.MutexRoot);
+                // avoid the RunInsideMutex to be nested used with same mutex name
+                t_innerCall.Value = t_innerCall.Value ?? new HashSet<string>();
+                if (!t_innerCall.Value.Add(lockPath))
+                {
+                    throw new ApplicationException($"Nested call to RunInsideMutex is detected, mutex name: {mutexName}");
```

Comment:
nit: moving this line inside `try` will cause `finally` to remove `lockPath` in case of exception.

Reply:
it's ok, remove just return false if the lockPath can't be found.